presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/InlineProjections.java

```diff
@@ -0,0 +1,174 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner.iterative.rule;
+
+import com.facebook.presto.Session;
+import com.facebook.presto.sql.planner.DependencyExtractor;
+import com.facebook.presto.sql.planner.PlanNodeIdAllocator;
+import com.facebook.presto.sql.planner.Symbol;
+import com.facebook.presto.sql.planner.SymbolAllocator;
+import com.facebook.presto.sql.planner.iterative.Lookup;
+import com.facebook.presto.sql.planner.iterative.Rule;
+import com.facebook.presto.sql.planner.plan.Assignments;
+import com.facebook.presto.sql.planner.plan.PlanNode;
+import com.facebook.presto.sql.planner.plan.ProjectNode;
+import com.facebook.presto.sql.tree.Expression;
+import com.facebook.presto.sql.tree.ExpressionRewriter;
+import com.facebook.presto.sql.tree.ExpressionTreeRewriter;
+import com.facebook.presto.sql.tree.Literal;
+import com.facebook.presto.sql.tree.SymbolReference;
+import com.facebook.presto.sql.tree.TryExpression;
+import com.facebook.presto.sql.util.AstUtils;
+import com.google.common.collect.Sets;
+
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import static java.util.stream.Collectors.toSet;
+
+/**
+ * Inlines expressions from a child project node into a parent project node
+ * as long as they are simple constants, or they referenced only once (to avoid
+ * introducing duplicate computation) and the references don't appear within
+ * a TRY block (to avoid changing semantics).
+ */
+public class InlineProjections
+        implements Rule
+{
+    @Override
+    public Optional<PlanNode> apply(PlanNode node, Lookup lookup, PlanNodeIdAllocator idAllocator, SymbolAllocator symbolAllocator, Session session)
+    {
+        if (!(node instanceof ProjectNode)) {
+            return Optional.empty();
+        }
+
+        ProjectNode parent = (ProjectNode) node;
+
+        PlanNode source = lookup.resolve(parent.getSource());
+        if (!(source instanceof ProjectNode)) {
+            return Optional.empty();
+        }
+
+        ProjectNode child = (ProjectNode) source;
+
+        Sets.SetView<Symbol> targets = extractInliningTargets(parent, child);
+        if (targets.isEmpty()) {
+            return Optional.empty();
+        }
+
+        // inline the expressions
+        Assignments assignments = child.getAssignments().filter(targets::contains);
+        Map<Symbol, Expression> parentAssignments = parent.getAssignments()
+                .entrySet().stream()
+                .collect(Collectors.toMap(
```

Comment:
Maybe there should be a collector in `Assignments` that allows for key and value lambda like `Collectors.toMap`, but instead creating `Assignments`?